### ISSUE TITLE
Low runner "createBuffer" error on iOS

### ISSUE THREAD

**Issue by bflora2** [goldfire/howler.js#654]
I am receiving reports of the following error from the field: `null is not an object (evaluating 'e.ctx.createBuffer')`. The error occurs in Howler 2.0.1 (minified) when running the following code:

```
new Howl({
   src: [ 'https://cdn.jigsawexplorer.com/puzzles/resources/audio/applause.webm',
          'https://cdn.jigsawexplorer.com/puzzles/resources/audio/applause.mp3' ],
   volume: 0.3,
   onload: callback,
   onloaderror: errCallback
});
```

The error appears to be coming from line 263 in howler.js (unminified).

The error only occurs on iOS devices, and it occurs for much less than 1% of iOS devices. However, devices that generate the error tend to generate it again in future sessions. I have been seeing the error for a number of months (it is not new to v2.0.1). I cannot recreate the error on my iPad.

If you need more information then I can add code to gather additional data when the error occurs. Just let me know what you would need.

**Comment by bflora2** [comment]
Correction: I've just received the same error for an Android device. After checking again I found two other earlier reports of this error from Android devices. So, it is not just iOS as I had incorrectly stated earlier, but no PCs or Macs have reported this error.

**Comment by goldfire** [comment]
Is this the line you are saying the error is firing on?

    self._scratchBuffer = self.ctx.createBuffer(1, 1, 22050);

**Comment by bflora2** [comment]
Yes. Apparently, in rare situations on iOS and (even rarer) on Android, the value of `ctx` is null on that line.

**Comment by goldfire** [comment]
Are you 100% it is referencing that line? A few lines above it already checks if self.ctx is set, so I'm not sure how it could possibly be null at that line.

**Comment by bflora2** [comment]
Yes, I'm sure. That is the only line in the source that calls `ctx.createBuffer`.

Yes, there is a null check performed on `ctx` a half dozen lines before the error is thrown. But, below that, immediately before the error is thrown, there appears to be an opportunity to call `self.unload` which can set `ctx` to null.

**Comment by bflora2** [comment]
After placing some log traces in the Howler code I now see the cause of the problem: The call to `self.unload` (line 292, v2.0.2) sets `self.ctx` to null. `unload` then calls `setupAudioContext` which normally assigns a new audio context to `self.ctx`, but in rare cases `setupAudioContext` will throw an error from native code due to a lack of audio resources. That means `self.ctx` is left set to null when `unload` returns, which in turn, causes `self._scratchBuffer = self.ctx.createBuffer(1, 1, 22050)` (line 297, v2.0.2) to throw an unhandled error.

It appears the problem can be fixed by adding a null check after the call to `self.unload` (line 292, v2.0.2):

```
if (!self._mobileUnloaded && self.ctx.sampleRate !== 44100) {
  self._mobileUnloaded = true;
  self.unload();

  if (!self.ctx) {
    return;
  }
}
```

**Comment by llli0v0** [comment]
I met the same problem, finally you is how to solve

**Comment by llli0v0** [comment]
I have a little don't understand, if Howler.usingWebAudio = false, what is self.ctx?

**Comment by AmirL** [comment]
BTW, The bug is still there :(.

**Comment by lbarthon** [comment]
Yes, fixed with #1363, waiting for a merge

**Comment by Ch1pStar** [comment]
Hello, we're currently experiencing this error in some very specific situations on iOS Safari, but the ```ctx``` object is properly created and is not null. The exception is coming from the ```createBuffer`` call, so checking if ctx is null doesn't fix the issue. As a temp fix I'm setting ```self.noAudio=true``` and returning, but hoping to find a proper fix for this.

**Comment by hmoffatt** [comment]
I am getting lots of null errors from inside the unlock function, different to the error above.
null is not an object (evaluating 'e.ctx.createBufferSource')

somehow ctx is valid when the touch handlers are added but not later when executed.

**Comment by lbarthon** [comment]
Maybe an up here? My pr has been opened for half a year now..

**Comment by 0xdevalias** [comment]
We're also seeing this issue pop up sporadically, would be awesome #1363 could get merged.

@goldfire and/or any other contributors able to weigh in here?

![image](https://user-images.githubusercontent.com/753891/113638969-7c190f80-96bb-11eb-8913-c86631f18c51.png)
![image](https://user-images.githubusercontent.com/753891/113638985-863b0e00-96bb-11eb-8ee4-4fad6a6f5f7e.png)

---

While it may not be related to this, when googling for the error I stumbled across the following, which sounds similar (though probably not as useful as the explicit tracing done in https://github.com/goldfire/howler.js/issues/654#issuecomment-266774785):

- https://github.com/goldfire/howler.js/issues/1021

**Comment by lbarthon** [comment]
<img src="https://i.barthonet.ovh/uploads/JPk9H.gif" />

**Comment by hmoffatt** [comment]
@0xdevalias I also see the suspend issue. Always Safari.
